atlite/resource.py

```diff
@@ -36,90 +36,252 @@
-    # unit is m^2 here
-
-    if isinstance(panel, str):
-        panel = get_solarpanelconfig(panel)
-
-    model = panel.get('model', 'huld')
-    if model == 'huld':
-        return panel['efficiency']
-    elif model == 'bofinger':
-        # one unit in the capacity layout is interpreted as one panel of a
-        # capacity (A + 1000 * B + log(1000) * C) * 1000W/m^2 * (k / 1000)
-        A, B, C = itemgetter('A', 'B', 'C')(panel)
-        return (A + B * 1000. + C * np.log(1000.))*1e3
-
-def windturbine_rated_capacity_per_unit(turbine):
-    if isinstance(turbine, str):
-        turbine = get_windturbineconfig(turbine)
+# Global caches
+_oedb_turbines = None
 
-    return turbine['P']
+class Resources:
+    def __init__(self, config):
+        self.config = config
+
+        self.windturbines = arrowdict()
+        self.solarpanels = arrowdict()
+
+        config.register_update_hook(self._update_resource_dictionaries)
+        self._update_resource_dictionaries(None)
+
+    def _update_resource_dictionaries(self, _):
+        self.windturbines.clear()
+        self.windturbines.update({p.stem: p.stem
+                                  for p in self.config.windturbine_dir.glob("*.yaml")})
+
+        self.solarpanels.clear()
+        self.solarpanels.update({p.stem: p.stem
+                                 for p in self.config.solarpanel_dir.glob("*.yaml")})
+
+    def get_windturbineconfig(self, turbine):
+        """Load the wind 'turbine' configuration.
+
+        The configuration can either be one from local storage, then 'turbine' is
+        considered part of the file base name '<turbine>.yaml' in `config.windturbine_dir`.
+        Alternatively the configuration can be downloaded from the Open Energy Database (OEDB),
+        in which case 'turbine' is a dictionary used for selecting a turbine from the database.
+
+        Parameter
+        ---------
+        turbine : str|dict
+            Name of the local turbine file.
+            Alternatively a dict for selecting a turbine from the Open Energy
+            Database, in this case the key 'source' should be contained. For all
+            other key arguments to retrieve the matching turbine, see
+            `atlite.resource.Resources.download_windturbineconfig()` for details.
+        """
+
+        def read_windturbineconfig(turbine):
+            res_name = self.config.windturbine_dir / f"{turbine}.yaml"
+            with open(res_name, "r") as turbine_file:
+                return yaml.safe_load(turbine_file)
+
+        def as_turbineconfig(turbine):
+            if isinstance(turbine, str):
+                if turbine.startswith('oedb:'):
+                    turbine = turbine[len("oedb:"):]
+                    return self.download_windturbineconfig(turbine, store_locally=False)
+                else:
+                    return read_windturbineconfig(turbine)
+            elif isinstance(turbine, dict):
+                if turbine.get('source') is None:
```

Comment:
The source keyword now no longer allows to specify a config directly as a dict.

I.e. this:

`cutout.get_windturbineconfig({"V":[1,2,3], "P":[4,5,6], "HUB_HEIGHT":42})`

is no longer possible, as per default "oedb" is assumed or I could specify "local".
I'd like to either have a dedicated keyword (e.g. 'manual') or have let default (no 'source') be directly as a dict.

Reply:
How would this work in the original code:
https://github.com/PyPSA/atlite/blob/045cda6bac0a70df2932ade373607d5b5953d83b/atlite/resource.py#L63-L83

Reply:
You're right. It wouldn't have worked back then. No need to change it then either.